src/compiler/debug.ts

```diff
@@ -107,1064 +106,1166 @@ export interface LoggingHost {
-    export function assertEachNode(nodes: readonly Node[], test: ((node: Node) => boolean) | undefined, message?: string, stackCrawlMark?: AnyFunction): void;
-    export function assertEachNode(nodes: readonly Node[] | undefined, test: ((node: Node) => boolean) | undefined, message?: string, stackCrawlMark?: AnyFunction) {
-        if (shouldAssertFunction(AssertionLevel.Normal, "assertEachNode")) {
-            assert(
-                test === undefined || every(nodes, test),
-                message || "Unexpected node.",
-                () => `Node array did not pass test '${getFunctionName(test!)}'.`,
-                stackCrawlMark || assertEachNode);
-        }
+/** @internal */
+export function assertEachIsDefined<T extends Node>(value: NodeArray<T>, message?: string, stackCrawlMark?: AnyFunction): asserts value is NodeArray<T>;
+/** @internal */
+export function assertEachIsDefined<T>(value: readonly T[], message?: string, stackCrawlMark?: AnyFunction): asserts value is readonly NonNullable<T>[];
+export function assertEachIsDefined<T>(value: readonly T[], message?: string, stackCrawlMark?: AnyFunction) {
+    for (const v of value) {
+        assertIsDefined(v, message, stackCrawlMark || assertEachIsDefined);
     }
+}
 
-    export function assertNode<T extends Node, U extends T>(node: T | undefined, test: (node: T) => node is U, message?: string, stackCrawlMark?: AnyFunction): asserts node is U;
-    export function assertNode(node: Node | undefined, test: ((node: Node) => boolean) | undefined, message?: string, stackCrawlMark?: AnyFunction): void;
-    export function assertNode(node: Node | undefined, test: ((node: Node) => boolean) | undefined, message?: string, stackCrawlMark?: AnyFunction) {
-        if (shouldAssertFunction(AssertionLevel.Normal, "assertNode")) {
-            assert(
-                node !== undefined && (test === undefined || test(node)),
-                message || "Unexpected node.",
-                () => `Node ${formatSyntaxKind(node?.kind)} did not pass test '${getFunctionName(test!)}'.`,
-                stackCrawlMark || assertNode);
-        }
+/** @internal */
+export function checkEachDefined<T, A extends readonly T[]>(value: A, message?: string, stackCrawlMark?: AnyFunction): A {
+    assertEachIsDefined(value, message, stackCrawlMark || checkEachDefined);
+    return value;
+}
+
+/** @internal */
+export function assertNever(member: never, message = "Illegal value:", stackCrawlMark?: AnyFunction): never {
+    const detail = typeof member === "object" && hasProperty(member, "kind") && hasProperty(member, "pos") ? "SyntaxKind: " + formatSyntaxKind((member as Node).kind) : JSON.stringify(member);
+    return fail(`${message} ${detail}`, stackCrawlMark || assertNever);
+}
+
+/** @internal */

```

Comment:
Why the switch from an overloaded function to a variable?

Reply:
As far as I know, this is the only way for me to declare a mutable binding that has an overloaded signature. Overloads aren't allowed on variable declarations, leaving me with interface/object + signature notation.

If we didn't do this "replace asserts with noops" thing, I'd just write them as regular functions. Maybe I can try forcing these debug asserts to be enabled in tsc (where they're normally not on) and test it both ways.

That being said, we have way, way, way more debug asserts that _don't_ have this behavior, e.g. the ubiquitous `Debug.assert` and `Debug.checkDefined`, so maybe it's not worth it to have this pattern anymore either in the grand scheme of things.

Reply:
Is this "replace asserts with noops" thing new as of this PR? I don't recall us doing that before. Does it have a measurable perf benefit? If so, we could always have the public export call into a non-exported variable and have that be the thing that gets replaced.

Reply:
It's not new as of this PR; I believe you added it a few years ago in #36995.

I'll check the perf, because I'd love to get rid of this hack (and honestly assert levels entirely).

Reply:
Ah, my bad. I was looking for assignments to `Debug.assertEachChild` and the like. I'd completely forgotten about this.

Reply:
I have checked and there appears to be no performance difference between this PR and just always calling `shouldAssert`. That seems to imply it's better to just drop this abstraction, which I like.

I am also currently testing what it'd be like to not even call `shouldAssert` and just _always_ perform the assertions for these node-related functions; I'll update later when that's done.

Reply:
It does seem like there's a slight performance loss (maybe a percent) if I don't use shouldAssert and instead always assert with these helpers; I'll just simplify this by dropping the whole cache/noop mechanism entirely.

I can also send that as its own PR before this one, if preferred.